controllers/tykoas_controller.go

```diff
@@ -0,0 +1,324 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/TykTechnologies/tyk-operator/api/v1alpha1"
+	tykv1 "github.com/TykTechnologies/tyk-operator/api/v1alpha1"
+	tykv1alpha1 "github.com/TykTechnologies/tyk-operator/api/v1alpha1"
+	"github.com/TykTechnologies/tyk-operator/pkg/client/klient"
+	"github.com/TykTechnologies/tyk-operator/pkg/environment"
+	"github.com/TykTechnologies/tyk-operator/pkg/keys"
+	"github.com/buger/jsonparser"
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	util "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	TykOASConfigMapKey  = "spec.tykOAS.configmapRef.name"
+	TykOASExtenstionStr = "x-tyk-api-gateway"
+)
+
+// TykOASReconciler reconciles a TykOAS object
+type TykOASReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Env    environment.Env
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=tyk.tyk.io,resources=tykoas,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=tyk.tyk.io,resources=tykoas/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=tyk.tyk.io,resources=tykoas/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the TykOAS object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *TykOASReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("TykOAS", req.NamespacedName.String())
+
+	log.Info("Reconciling Tyk OAS instance")
+
+	var reqA time.Duration
+	var apiID string
+	var markForDeletion bool
+
+	// Lookup Tyk OAS object
+	tykOAS := &tykv1.TykOAS{}
+	if err := r.Get(ctx, req.NamespacedName, tykOAS); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	_, ctx, err := HttpContext(ctx, r.Client, &r.Env, tykOAS, log)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	_, err = util.CreateOrUpdate(ctx, r.Client, tykOAS, func() error {
+		if !tykOAS.ObjectMeta.DeletionTimestamp.IsZero() {
+			markForDeletion = true
+			err := r.delete(ctx, tykOAS)
+			if err != nil {
+				return err
+			}
+
+			util.RemoveFinalizer(tykOAS, keys.TykOASFinalizerName)
+
+			return nil
+		}
+
+		util.AddFinalizer(tykOAS, keys.TykOASFinalizerName)
+
+		apiID, err = r.createOrUpdateTykOASAPI(ctx, tykOAS)
+		if err != nil {
+			r.Log.Error(err, "Failed to create/update Tyk OAS API", "name", client.ObjectKeyFromObject(tykOAS))
```

Comment:
I guess we can use `log` here instead of `r.Log` as we have access to `log` 
```suggestion
			log.Error(err, "Failed to create/update Tyk OAS API")
```

Reply:
Resolved in [e3d178e](https://github.com/TykTechnologies/tyk-operator/pull/687/commits/e3d178ef016740ad1622181e3b43778641d34e37)